sotodlib/io/bookbinder2.py

```diff
@@ -0,0 +1,1022 @@
+        else:
+            self.log = log
+
+    def preprocess(self):
+        """
+        Extracts file times, nchans, and nnframes from file list
+        """
+        if self.times is not None:  # Already preprocessed
+            return
+
+        self.log.info(f"Preprocessing smurf obsid {self.obs_id}")
+
+        self.nframes = 0
+        ts = []
+        frame_idxs = []
+        frame_idx = 0
+        for frame in get_frame_iter(self.files):
+            if frame.type != core.G3FrameType.Scan:
+                continue
+
+
+            if self.nchans is None:
+                self.nchans = len(self.readout_ids)
+                self.primary_names = frame['primary'].names
+                self.bias_names = frame['tes_biases'].names
+                self.timing_paradigm = frame['timing_paradigm']
+                self.session_id = frame['session_id']
+                if 'slow_primary' in frame:
+                    self.slow_primary = frame['slow_primary']
+                self.sostream_version = frame['sostream_version']
+                self.stream_id = frame['sostream_id']
+
+            t = get_frame_times(frame)[1]
+            ts.append(t)
+            frame_idxs.append(np.full(len(t), frame_idx, dtype=np.int))
+
+            self.nframes += 1
+            frame_idx += 1
+
+        self.times = np.hstack(ts)
+        self.frame_idxs = np.hstack(frame_idxs)
+        self.preprocessed = True
+
+    def bind(self, outdir, times, frame_idxs, file_idxs, pbar=False, ancil=None,
+             atol=1e-4):
+        """
+        Binds SMuRF data
+        
+        Params
+        ------ 
+        outdir : str
+            Output directory to put bound files
+        times : np.ndarray
+            Full array of timestamps that should be contained in the books
+        frame_idxs : np.ndarray
+            Output frame idx for each specified timestamp
+        file_idxs : np.ndarray
+            Output file idx for each specified output frame
+        pbar : bool or tqdm.tqdm
+            If True, will create a new progress bar. If False, will disable.
+            If a progress bar is passed, will use that.
+        ancil : AncilProcessor
+            Ancillary processor object (must be already bound). Ancil data will
+            be copied into the output frames.
+        atol : float
+            Absolute tolerance between smurf-timestamps and book-timestamps.
+            Samples mapped to times that are further away than atol will
+            be considered unmapped.
+        """
+        if pbar is True:
+            pbar = tqdm(total=self.nframes)
+        elif pbar is False:
+            pbar = tqdm(total=self.nframes, disable=True)
+
+        pbar.set_description(f"Binding {self.stream_id}")
+
+        self.log.info(f"Binding smurf obsid {self.obs_id}")
+
+        # Mapping to input sample idx to output sample idx
+        atol = 1e-4
+        sample_map = find_ref_idxs(times, self.times)
```

Comment:
a comment indicating times has no gap, self.times have gaps will be helpful for reading